packages/story-editor/src/components/layout/index.js

```diff
@@ -81,24 +88,170 @@ function Layout({ header, children }) {
+    'web-stories'
+  );
+  if (allowedFileTypes.length) {
+    onSelectErrorMessage = sprintf(
+      /* translators: %s: list of allowed file types. */
+      __('Please choose only %s to insert into page.', 'web-stories'),
+      translateToExclusiveList(allowedFileTypes)
+    );
+  }
+
+  /**
+   * Insert element such image, video and audio into the editor.
+   *
+   * @param {Object} resource Resource object
+   * @param {string} thumbnailURL The thumbnail's url
+   * @return {null|*} Return onInsert or null.
+   */
+  const insertMediaElement = useCallback(
+    (resource, thumbnailURL) => {
+      resourceList.set(resource.id, {
+        url: thumbnailURL,
+        type: 'cached',
+      });
+      updateElementsById({
+        elementIds: [selectedElements?.[0]?.id],
+        properties: { resource },
+      });
+    },
+    [selectedElements, updateElementsById]
+  );
+
+  const handleMediaSelect = useCallback(
+    (resource) => {
+      try {
+        if (isTranscodingEnabled && canTranscodeResource(resource)) {
+          if (transcodableMimeTypes.includes(resource.mimeType)) {
+            optimizeVideo({ resource });
+          }
+
+          if (resource.mimeType === 'image/gif') {
+            optimizeGif({ resource });
+          }
+        }
+        // WordPress media picker event, sizes.medium.source_url is the smallest image
+        insertMediaElement(
+          resource,
+          resource.sizes?.medium?.source_url || resource.src
+        );
+
+        if (
+          !resource.local &&
+          allowedVideoMimeTypes.includes(resource.mimeType) &&
+          resource.isMuted === null
+        ) {
+          updateVideoIsMuted(resource.id, resource.src);
+        }
+      } catch (e) {
+        showSnackbar({
+          message: e.message,
+          dismissable: true,
+        });
+      }
+    },
+    [
+      allowedVideoMimeTypes,
+      insertMediaElement,
+      isTranscodingEnabled,
+      optimizeGif,
+      optimizeVideo,
+      showSnackbar,
+      transcodableMimeTypes,
+      updateVideoIsMuted,
+    ]
+  );
+
   return (
     <>
       <LayoutProvider>
         <ChecklistCheckpointProvider>
-          <HighlightsProvider>
-            <Editor zIndex={3}>
-              <CanvasProvider>
-                <RightClickMenuProvider>
-                  <Area area="lib">
-                    <Library />
-                  </Area>
-                  <Workspace header={header} />
-                  <RightClickMenu />
-                </RightClickMenuProvider>
-              </CanvasProvider>
-              {children}
-            </Editor>
-          </HighlightsProvider>
+          <MediaUploadProvider
+            title={__('Replace media', 'web-stories')}
+            buttonInsertText={__('Insert media', 'web-stories')}
+            onSelect={handleMediaSelect}
+            onClose={resetWithFetch}
+            type={allowedMimeTypes}
+            onSelectErrorMessage={onSelectErrorMessage}
+          >
```

Comment:
That's quite odd, as these props would mean this generically named `MediaUploadProvider` is only used for the "Replace media" use case.

Shouldn't these be passed to `onOpenMediaUpload` if anything?

Reply:
I think that would be tough, cus you'd need to make sure the `MediaUpload` rendered with those args before the `open` it supplies (and you'd be calling) is valid. I think this is just one of the caveats that comes with making this a component that exposes `open` through a render prop.

Kinda mentioned it here:
https://github.com/google/web-stories-wp/issues/8953#issuecomment-955018168

Reply:
In this case, it is only for the 'replace media' use case. For any other use cases we'd have to create another instance of the provider.

Just like Max said - this is part of having the `MediaUpload` component passed into the config, and not exposing an `openModal` function without first rendering the component. This provider exposes the 'open' function to the rest of the app here: https://github.com/google/web-stories-wp/pull/9534/files#diff-fb9f988abfb4a21e672cc1f4d632c65fe4815ea00db7a5936443cd0b04e83e2eR91-R98.

The problem here is that the 'open' function can't be created until we pass those props into the component.

Reply:
Then at the very least this provider would need to be renamed to have a less generic name